pkg/http/listener.go

```diff
@@ -0,0 +1,326 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package http
+
+import (
+	"crypto/tls"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+)
+
+var sslRequiredErrMsg = []byte("HTTP/1.0 403 Forbidden\r\n\r\nSSL required")
+
+// HTTP methods.
+var methods = []string{
+	http.MethodGet,
+	http.MethodHead,
+	http.MethodPost,
+	http.MethodPut,
+	http.MethodPatch,
+	http.MethodDelete,
+	http.MethodConnect,
+	http.MethodOptions,
+	http.MethodTrace,
+	"PRI", // HTTP 2 method
+}
+
+// maximum length of above methods + one space.
+var methodMaxLen = getMethodMaxLen() + 1
+
+func getMethodMaxLen() int {
+	maxLen := 0
+	for _, method := range methods {
+		if len(method) > maxLen {
+			maxLen = len(method)
+		}
+	}
+
+	return maxLen
+}
+
+func isHTTPMethod(s string) bool {
+	for _, method := range methods {
+		if s == method {
+			return true
+		}
+	}
+
+	return false
+}
+
+type acceptResult struct {
+	conn net.Conn
+	err  error
+}
+
+// httpListener - HTTP listener capable of handling multiple server addresses.
+type httpListener struct {
+	mutex                  sync.Mutex         // to guard Close() method.
+	tcpListeners           []*net.TCPListener // underlaying TCP listeners.
+	acceptCh               chan acceptResult  // channel where all TCP listeners write accepted connection.
+	doneChs                []chan struct{}    // done channels for each goroutine for TCP listeners.
+	tlsConfig              *tls.Config        // TLS configuration
+	tcpKeepAliveTimeout    time.Duration
+	readTimeout            time.Duration
+	writeTimeout           time.Duration
+	updateBytesReadFunc    func(int)                           // function to be called to update bytes read in bufConn.
+	updateBytesWrittenFunc func(int)                           // function to be called to update bytes written in bufConn.
+	errorLogFunc           func(error, string, ...interface{}) // function to be called on errors.
+}
+
+// start - starts separate goroutine for each TCP listener.  A valid insecure/TLS HTTP new connection is passed to httpListener.acceptCh.
+func (listener *httpListener) start() {
+	listener.acceptCh = make(chan acceptResult)
+
+	// Closure to send acceptResult to acceptCh.
+	// It returns true if the result is sent else false if returns when doneCh is closed.
+	send := func(result acceptResult, doneCh <-chan struct{}) bool {
+		select {
+		case listener.acceptCh <- result:
+			// Successfully written to acceptCh
+			return true
+		case <-doneCh:
+			// As stop signal is received, close accepted connection.
+			if result.conn != nil {
+				result.conn.Close()
+			}
+			return false
+		}
+	}
+
+	// Closure to accept single connection.
+	acceptTCP := func(tcpConn *net.TCPConn, doneCh <-chan struct{}) {
```

Comment:
If they are methods to `httpListener`, they should be called only by `start()` ie they have no meaning outside `start()` and there is no control on whether they are called only by `start()`.  Secondly they can't be unit tested.